src/charm.py

```diff
@@ -537,22 +547,62 @@ def _configure_pki_secrets_engine(self) -> None:  # noqa: C901
             private_key=str(private_key),
             mount=PKI_MOUNT,
         )
+        issued_certificates_validity = self._calculate_pki_certificates_validity(
+            provider_certificate.certificate
+        )
         if not vault.is_common_name_allowed_in_pki_role(
             role=PKI_ROLE_NAME,
             mount=PKI_MOUNT,
             common_name=config_common_name,
+        ) or issued_certificates_validity != vault.get_role_max_ttl(
+            role=PKI_ROLE_NAME, mount=PKI_MOUNT
         ):
             vault.create_or_update_pki_charm_role(
                 allowed_domains=config_common_name,
                 mount=PKI_MOUNT,
                 role=PKI_ROLE_NAME,
+                max_ttl=f"{issued_certificates_validity}s",
             )
         # Can run only after the first issuer has been actually created.
         try:
             vault.make_latest_pki_issuer_default(mount=PKI_MOUNT)
         except VaultClientError as e:
             logger.error("Failed to make latest issuer default: %s", e)
 
+    def _intermediate_ca_is_active(
+        self, vault: Vault, intermediate_ca_certificate: Certificate
+    ) -> bool:
+        """Check if intermediate CA can sign certificates.
+
+        Vault PKI doesn't allow signing certificates that would outlast the CA.
+        We check here that the time remaining for the CA to expire is longer than the
+        certificates we are about to issue.
+        """
+        current_ttl = vault.get_role_max_ttl(role=PKI_ROLE_NAME, mount=PKI_MOUNT)
+        if (
+            not current_ttl
+            or not intermediate_ca_certificate.expiry_time
+            or not intermediate_ca_certificate.validity_start_time
+        ):
+            return False
+        certificate_validity = (
+            intermediate_ca_certificate.expiry_time
+            - intermediate_ca_certificate.validity_start_time
+        )
+        certificate_validity_seconds = certificate_validity.total_seconds()
+        return certificate_validity_seconds > current_ttl
+
+    def _calculate_pki_certificates_validity(self, certificate: Certificate) -> int:
+        """Calculate the maximum allowed validity of certificates issued by PKI.
+
+        Return half the CA certificate validity.
```

Comment:
We should likely document this in our documentation (after this is merged). It should be quite clear that "when using Vault as an intermediate CA authority, its own certificate validity will be determined by the TLS providers and the certificates it issues will be valid for half this time."

Reply:
Yes, I agree.